### PR TITLE
feat(sync): `TxLookup` stage

### LINKED ISSUE #969: Transaction Lookup Stage
### Describe the feature

Move writing to `TxHashNumber` table out of the bodies stage into a separate `TxLookup` stage.
https://github.com/paradigmxyz/reth/blob/583426bb95cfa65d5f5e1ee29e084f34620e5813/crates/stages/src/stages/bodies.rs#L141

This stage would collect all of the transactions from the input block range and insert entries into the `TxHashNumber` table.

### Additional context

The `Bodies` stage wastes a lot of time writing to `TxHashNumber` table. Since it's blocked on the writes to db, it cannot advance network requests further.

### PR DESCRIPTION
**Pull request by gakonst**
Closes #969 

This PR introduces a new `TransactionLookup` stage for staged sync.

Previously, `TxHashNumber` table for storing the mapping of transaction hash to transaction id was populated during the `Bodies` stage. Since the writes to this table are unsorted, they cannot be optimized neither with appends nor with cursor inserts. This introduced a major delay to a body download.

The delay is mitigated by extracting the writes to a separate stage.

### CODE REVIEW

File: crates/stages/src/stages/tx_lookup.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        input: ExecInput,
+    ) -> Result<ExecOutput, StageError> {
+        let ((start_block, end_block), capped) =
+            exec_or_return!(input, self.commit_threshold, "sync::stages::transaction_lookup");
+
+        debug!(target: "sync::stages::transaction_lookup", start_block, end_block, "Commencing sync");
+
+        let mut cursor_bodies = tx.cursor_read::<tables::BlockBodies>()?;
+        let mut tx_cursor = tx.cursor_write::<tables::Transactions>()?;
+        let (mut current_tx_id, _) = tx.get_next_block_ids(start_block)?;
```
Comment (rkrasiuk) [inline crates/stages/src/stages/tx_lookup.rs:46]: this is incorrect (it gets tx id for `start_block + 1`) and redundant since you have ids from `body.tx_id_range()`

**Review by rkrasiuk — CHANGES_REQUESTED** [review]

File: crates/stages/src/stages/bodies.rs [hunk 2]
```diff
@@ -211,12 +206,9 @@ impl<DB: Database, D: BodyDownloader, C: Consensus> Stage<DB> for BodyStage<D, C
 
             // Delete all transactions that belong to this block
             for tx_id in body.tx_id_range() {
-                // First delete the transaction and hash to id mapping
-                if let Some((_, transaction)) = transaction_cursor.seek_exact(tx_id)? {
+                // First delete the transaction
+                if let Some((_, _)) = transaction_cursor.seek_exact(tx_id)? {
```
Comment (mattsse) [inline crates/stages/src/stages/bodies.rs:210]: sam as `is_some()`?

File: crates/stages/src/stages/bodies.rs [hunk 3]
```diff
@@ -453,9 +445,8 @@ mod tests {
             .tx()
             .commit(|tx| {
                 let mut tx_cursor = tx.cursor_write::<tables::Transactions>()?;
-                let (_, transaction) = tx_cursor.last()?.expect("Could not read last transaction");
+                tx_cursor.last()?.expect("Could not read last transaction");
```
Comment (mattsse) [inline crates/stages/src/stages/bodies.rs:448]: this is an unrelated q, under what circumstances would this panic?

File: crates/stages/src/stages/bodies.rs [hunk 4]
```diff
@@ -748,11 +733,6 @@ mod tests {
                                 tx_transition_cursor.seek_exact(tx_id).expect("to be okay").expect("to be present").1, current_transition_id
                             );
                             current_transition_id += 1;
```
Comment (mattsse) [inline crates/stages/src/stages/bodies.rs:735]: We have "+ 1" very often in the phases, and it is not always easy to understand why.
I'd like to a note for every `+1`

perhaps this is not applicable here. not sure what the `current_transition_id` is used for, but for example in headers stage we have this a lot and it's not well documented.

File: crates/stages/src/stages/sender_recovery.rs [hunk 5]
```diff
@@ -151,18 +150,13 @@ mod tests {
...
         // Insert blocks with a single transaction at block `stage_progress + 10`
-        (stage_progress..input.previous_stage_progress() + 1)
-            .map(|number| -> Result<SealedBlock, TestRunnerError> {
-                let tx_count = Some((number == stage_progress + 10) as u8);
-                let block = random_block(number, None, tx_count, None);
-                current_tx_id = runner.insert_block(current_tx_id, &block, false)?;
-                Ok(block)
+        let blocks = (stage_progress..input.previous_stage_progress() + 1)
+            .map(|number| {
+                random_block(number, None, Some((number == stage_progress + 10) as u8), None)
```
Comment (mattsse) [inline crates/stages/src/stages/sender_recovery.rs:156]: why `10`?

File: crates/stages/src/stages/tx_lookup.rs [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+                    .map(|(block_num_hash, _)| block_num_hash.number() <= end_block)
+                    .unwrap_or_default()
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        for (_, body) in bodies {
+            let transactions = tx_cursor
+                .walk(body.start_tx_id)?
+                .take(body.tx_count as usize)
+                .collect::<Result<Vec<_>, _>>()?;
```
Comment (mattsse) [inline crates/stages/src/stages/tx_lookup.rs:62]: please add brief explainer of what this does

File: crates/stages/src/stages/tx_lookup.rs [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    transaction::{DbTx, DbTxMut},
+};
+use tracing::*;
+
+const TRANSACTION_LOOKUP: StageId = StageId("TransactionLookup");
+
+/// The transaction lookup stage.
+///
+/// This stage walks over the bodies table, and sets the transaction hash of each transaction in a
+/// block to the correesponding `TransitionId` at each block. This is written to the
```
Comment (mattsse) [inline crates/stages/src/stages/tx_lookup.rs:18]:
```suggestion
/// block to the corresponding `TransitionId` at each block. This is written to the
```

File: crates/stages/src/test_utils/test_db.rs [hunk 8]
```diff
@@ -183,4 +183,44 @@ impl TestTransaction {
...
+
+                // Insert into body tables.
+                tx.put::<tables::BlockBodies>(
+                    key,
+                    StoredBlockBody {
+                        start_tx_id: current_tx_id,
+                        tx_count: block.body.len() as u64,
+                    },
+                )?;
+                for body_tx in block.body.clone() {
```
Comment (mattsse) [inline crates/stages/src/test_utils/test_db.rs:217]: is this clone required? seems expensive

File: crates/stages/src/test_utils/test_db.rs [hunk 9]
```diff
@@ -183,4 +183,44 @@ impl TestTransaction {
...
+        &self,
+        blocks: I,
+        tx_offset: Option<u64>,
+    ) -> Result<(), DbError>
+    where
+        I: Iterator<Item = &'a SealedBlock>,
+    {
+        self.commit(|tx| {
+            let mut current_tx_id = tx_offset.unwrap_or_default();
+            let blocks = blocks.collect::<Vec<_>>();
```
Comment (mattsse) [inline crates/stages/src/test_utils/test_db.rs:199]: is the collect redundant?

**Review by mattsse — CHANGES_REQUESTED** [review]
a few suggestions, nits and questions

File: crates/stages/src/stages/sender_recovery.rs (hunk 5 above)
Comment (rkrasiuk) [reply crates/stages/src/stages/sender_recovery.rs:156]: arbitrary number for a non-empty block

File: crates/stages/src/test_utils/test_db.rs (hunk 9 above)
Comment (rkrasiuk) [reply crates/stages/src/test_utils/test_db.rs:199]: yup, good catch

File: crates/stages/src/test_utils/test_db.rs (hunk 8 above)
Comment (rkrasiuk) [reply crates/stages/src/test_utils/test_db.rs:217]: yeah, because we are iterating over references

File: crates/stages/src/stages/bodies.rs (hunk 3 above)
Comment (rkrasiuk) [reply crates/stages/src/stages/bodies.rs:448]: i'd say if the test was written incorrectly

File: crates/stages/src/stages/bodies.rs (hunk 4 above)
Comment (rkrasiuk) [reply crates/stages/src/stages/bodies.rs]: renamed and left some comments. should be more clear now

File: crates/stages/src/stages/tx_lookup.rs (hunk 6 above)
Comment (rkrasiuk) [reply crates/stages/src/stages/tx_lookup.rs]: done

File: crates/stages/src/stages/sender_recovery.rs [hunk 10]
```diff
@@ -151,18 +150,14 @@ mod tests {
...
-                current_tx_id = runner.insert_block(current_tx_id, &block, false)?;
-                Ok(block)
+        let non_empty_block_number = stage_progress + 10;
+        let blocks = (stage_progress..input.previous_stage_progress() + 1)
+            .map(|number| {
+                random_block(number, None, Some((number == non_empty_block_number) as u8), None)
             })
-            .collect::<Result<Vec<_>, _>>()
-            .expect("failed to insert blocks");
+            .collect::<Vec<_>>();
```
Comment (mattsse) [inline crates/stages/src/stages/sender_recovery.rs:159]: do we need to collect here if we're only using `blocks.iter()`?

File: crates/stages/src/stages/sender_recovery.rs [hunk 11]
```diff
@@ -232,6 +227,22 @@ mod tests {
         fn set_threshold(&mut self, threshold: u64) {
             self.threshold = threshold;
         }
+
+        fn check_no_senders_by_block(&self, block: BlockNumber) -> Result<(), TestRunnerError> {
```
Comment (mattsse) [inline crates/stages/src/stages/sender_recovery.rs:231]: naming nit: 
```suggestion
        fn ensure_no_senders_by_block(&self, block: BlockNumber) -> Result<(), TestRunnerError> {
```

File: crates/stages/src/stages/sender_recovery.rs [hunk 12]
```diff
@@ -232,6 +227,22 @@ mod tests {
...
+            let body_result = self.tx.inner().get_block_body_by_num(block);
+            match body_result {
+                Ok(body) => self
+                    .tx
+                    .check_no_entry_above::<tables::TxSenders, _>(body.last_tx_index(), |key| {
```
Comment (mattsse) [inline crates/stages/src/stages/sender_recovery.rs:236]: same nit here, but can be handled separately

File: crates/stages/src/stages/sender_recovery.rs (hunk 11 above)
Comment (mattsse) [inline crates/stages/src/stages/sender_recovery.rs:231]: one line doc with `# Panics` section

File: crates/stages/src/stages/tx_lookup.rs [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+/// The transaction lookup stage.
+///
+/// This stage walks over the bodies table, and sets the transaction hash of each transaction in a
+/// block to the corresponding `TransitionId` at each block. This is written to the
+/// [`tables::TxHashNumber`] This is used for looking up changesets via the transaction hash.
+#[derive(Debug)]
+pub struct TransactionLookupStage {
+    /// The number of table entries to commit at once
+    pub commit_threshold: u64,
+}
```
Comment (mattsse) [inline crates/stages/src/stages/tx_lookup.rs:24]: nit: would rather have a `new` function instead of init struct via public, so `StageBuilder` will be consistent with how it creates the stages

File: crates/stages/src/stages/tx_lookup.rs [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+        // Walk over block bodies within a specified range.
+        let bodies = cursor_bodies
+            .walk(start_key)?
+            .take_while(|entry| {
+                entry
+                    .as_ref()
+                    .map(|(block_num_hash, _)| block_num_hash.number() <= end_block)
+                    .unwrap_or_default()
+            })
+            .collect::<Result<Vec<_>, _>>()?;
```
Comment (mattsse) [inline crates/stages/src/stages/tx_lookup.rs:57]: is this collect required because of the subsequent `tx_cursor.walk` over transactions?
in any case, looks like we only need the `start_tx_id` and `tx_count`

File: crates/stages/src/stages/tx_lookup.rs [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+
+        // Collect tranasctions for each body and insert the reverse lookup for hash -> tx_id.
+        for (_, body) in bodies {
+            let transactions = tx_cursor
+                .walk(body.start_tx_id)?
+                .take(body.tx_count as usize)
+                .collect::<Result<Vec<_>, _>>()?;
+
+            for (id, transaction) in transactions {
```
Comment (mattsse) [inline crates/stages/src/stages/tx_lookup.rs:66]: I guess we can use the transactions iter here directly and don't need to collect?

File: crates/stages/src/stages/sender_recovery.rs (hunk 10 above)
Comment (mattsse) [reply crates/stages/src/stages/sender_recovery.rs:159]: just realized I already asked this

**Review by mattsse — APPROVED** [review]
